modules/command-registry.js

```diff
@@ -0,0 +1,402 @@
+ */
+const _removeCommand = (commandName) => {
+    if (StringUtils.isEmpty(commandName)) {
+        return;
+    }
+
+    program.commands = program.commands.filter(
+        (registeredCommand) =>
+            !_commandEqualsByName(registeredCommand, commandName)
+    );
+};
+
+/**
+ * Sorts the command list alphabetically by name
+ */
+const _sortCommandsByName = () => {
+    program.commands = program.commands.sort(
+        (commandA, commandB) =>
+            commandA.name().toLowerCase() > commandB.name().toLowerCase()
+    );
+};
+
+/**
+ * Validates that the given command name has a value, and exists in the base command definition array,
+ * returning the found definition (or exiting if provided invalid input)
+ *
+ * @param {string} commandName
+ * @returns {CommandDefinition} The CommandDefinition for the requested command
+ */
+const _validateAndGetBaseCommand = (commandName) => {
+    const commandNames = BASE_COMMAND_DEFINITIONS.map(
+        (commandDefinition) => commandDefinition.command
+    ).join(", ");
+
+    if (StringUtils.isEmpty(commandName)) {
+        echo.error(
+            `Command name is required. Available commands are: ${commandNames}`
+        );
+        shell.exit(1);
+        // Returning here for the purpose of testing, shell.exit will kill the process in a normal run
+        return;
+    }
+
+    const baseCommand = BASE_COMMAND_DEFINITIONS.find(
+        (commandDefinition) =>
+            commandDefinition.command.toLowerCase() ===
+            commandName.toLowerCase()
+    );
+
+    if (baseCommand == null) {
+        echo.error(
+            `The specified command '${commandName}' was not found. Available commands are: ${commandNames}`
+        );
+        shell.exit(1);
+        // Returning here for the purpose of testing, shell.exit will kill the process in a normal run
+        return;
+    }
+
+    return baseCommand;
+};
+
+/**
+ * Validates a command registration based on whether a command of the same name is already registered,
+ * and if the consumer has opted in to overriding it.
+ *
+ * If the command exists and the user has opted in to overriding it, it will remove the original command.
+ *
+ * @param {CommandDefinition} commandDefinition
+ * @param {boolean} overrideIfRegistered
+ */
+const _validateCommandRegistration = (
+    commandDefinition,
+    overrideIfRegistered = false
+) => {
+    const { command } = commandDefinition;
+
+    // First, check to see if command is already registered.
+    const commandIsRegistered = _getCommand(command) != null;
+
+    if (commandIsRegistered && !overrideIfRegistered) {
+        echo.warn(
+            `Command '${command}' has already been registered and 'overrideIfRegistered' is set to false. Skipping this command registration.`
+        );
+        return false;
+    }
+
+    if (commandIsRegistered && overrideIfRegistered) {
+        // Filter out the registered command of the same name
+        _removeCommand(command);
```

Comment:
Should removal be handled in this function? I feel like this function should simply handle validation (is the command already registered?) and then the caller will do what it will with that info (remove and re-register, ignore, etc.). To me, calling a `validate` command would simply do validation, not perform additional functions like removing said command if I want to override it.

Reply:
That's fair 😛 Went through several refactors to get here, but it probably does make sense to separate it out and let the caller handle it. I'll update this.